tests/src/junit/java/net/neoforged/neoforge/unittest/transfer/ItemTests.java

```diff
@@ -0,0 +1,261 @@
+            try (Transaction transaction = Transaction.open(null)) {
+                // Insert bucket from down - should fail.
+                if (downWrapper.insert(emptyBucket, 1, transaction) != 0) throw new AssertionError("Bucket should not have been inserted.");
+                // Insert bucket unsided - should go in slot 1 (canPlaceItem returns false for slot 0).
+                if (unsidedWrapper.insert(emptyBucket, 1, transaction) != 1) throw new AssertionError("Failed to insert bucket.");
+                if (!testContainer.getItem(0).isEmpty()) throw new AssertionError("Slot 0 should have been empty.");
+                if (!stackEquals(testContainer.getItem(1), Items.BUCKET, 1)) throw new AssertionError("Slot 1 should have been a bucket.");
+                // The bucket should be extractable from any side but the top.
+                if (!emptyBucket.equals(ResourceHandlerUtil.findExtractableResource(unsidedWrapper, r -> true, transaction))) throw new AssertionError("Bucket should be extractable from unsided wrapper.");
+                if (!emptyBucket.equals(ResourceHandlerUtil.findExtractableResource(downWrapper, r -> true, transaction))) throw new AssertionError("Bucket should be extractable from down wrapper.");
+                if (ResourceHandlerUtil.findExtractableResource(upWrapper, r -> true, transaction) != null) throw new AssertionError("Bucket should NOT be extractable from up wrapper.");
+
+                if (iter == 1) {
+                    // Commit the second time only.
+                    transaction.commit();
+                }
+            }
+        }
+
+        // Check commit.
+        if (!testContainer.getItem(0).isEmpty()) throw new AssertionError("Slot 0 should have been empty.");
+        if (!testContainer.getItem(1).is(Items.BUCKET) || testContainer.getItem(1).getCount() != 1) throw new AssertionError("Slot 1 should have been a bucket.");
+
+        checkComparatorOutput(testContainer);
+
+        // Check that we return sensible results if amount stored > capacity
+        ItemStack oversizedStack = new ItemStack(Items.DIAMOND_PICKAXE, 2);
+        SimpleContainer simpleContainer = new SimpleContainer(oversizedStack);
+        var wrapper = VanillaContainerWrapper.of(simpleContainer);
+
+        try (Transaction transaction = Transaction.open(null)) {
+            assertEquals(0L, wrapper.insert(ItemResource.of(oversizedStack), 10, transaction));
+            transaction.commit();
+        }
+    }
+
+    private static boolean stackEquals(ItemStack stack, Item item, int count) {
+        return stackEquals(stack, ItemResource.of(item), count);
+    }
+
+    private static boolean stackEquals(ItemStack stack, ItemResource variant, int count) {
+        return variant.matches(stack) && stack.getCount() == count;
+    }
+
+    private static class TestWorldlyContainer extends SimpleContainer implements WorldlyContainer {
+        private static final int[] SLOTS = new int[] { 0, 1, 2 };
```

Comment:
Unnecessary explicit `new int[]` (for that matter most of the `public`'s on test methods are also unnecessary for junit